src/NHibernate.Test/SecondLevelCacheTest/NeverCachedEntityTests.cs

```diff
@@ -0,0 +1,170 @@
+
+		protected override string[] Mappings => new[] { "SecondLevelCacheTest.Item.hbm.xml" };
+
+		protected override void Configure(Configuration configuration)
+		{
+			configuration.SetProperty(Environment.CacheProvider, typeof(HashtableCacheProvider).AssemblyQualifiedName);
+			configuration.SetProperty(Environment.UseQueryCache, "true");
+		}
+
+		[Test]
+		public void NeverInvalidateEntities()
+		{
+			var debugSessionFactory = (DebugSessionFactory) Sfi;
+
+			var cache = Substitute.For<UpdateTimestampsCache>(Sfi.Settings, new Dictionary<string, string>());
+
+			var updateTimestampsCacheField = typeof(SessionFactoryImpl).GetField(
+				"updateTimestampsCache",
+				BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
+
+			updateTimestampsCacheField.SetValue(debugSessionFactory.ActualFactory, cache);
+
+			//"Received" assertions can not be used since the collection is reused and cleared between calls.
+			//The received args are cloned and stored
+			var preInvalidations = new List<IReadOnlyCollection<string>>();
+			var invalidations = new List<IReadOnlyCollection<string>>();
+
+			cache.PreInvalidate(Arg.Do<IReadOnlyCollection<string>>(x => preInvalidations.Add(x.ToList())));
+			cache.Invalidate(Arg.Do<IReadOnlyCollection<string>>(x => invalidations.Add(x.ToList())));
+
+			using (var session = OpenSession())
+			{
+				//Add NeverItem
+				using (var tx = session.BeginTransaction())
+				{
+					foreach (var i in Enumerable.Range(1, 10))
+					{
+						var item = new NeverItem 
+						{
+							Id = i, Name="NHibernate"+i,
+							Description="Cache.Never() triggers UpdateTimestampsCache" 
+						};
+						session.Save(item);
+					}
+
+					tx.Commit();
+				}
+
+				//Update NeverItem
+				using (var tx = session.BeginTransaction())
+				{
+					foreach (var i in Enumerable.Range(1, 10))
+					{
+						var item = session.Get<NeverItem>(i);
+						item.Name = item.Id.ToString();
+					}
+
+					tx.Commit();
+				}
+
+				//Delete NeverItem
+				using (var tx = session.BeginTransaction())
+				{
+					foreach (var i in Enumerable.Range(1, 10))
+					{
+						var item = session.Get<NeverItem>(i);
+						session.Delete(item);
+					}
+
+					tx.Commit();
+				}
+
+				//Update NeverItem using HQL
+				using (var tx = session.BeginTransaction())
+				{
+					session.CreateQuery("UPDATE NeverItem SET Name='Test'").ExecuteUpdate();
+
+					tx.Commit();
+				}
+
+				//Update NeverItem using LINQ
+				using (var tx = session.BeginTransaction())
+				{
+					session.Query<NeverItem>()
+						   .UpdateBuilder()
+						   .Set(x => x.Name, "Test")
+						   .Update();
+
+					tx.Commit();
+				}
+			}
+
+			//Should receive none preinvalidation when Cache is configured as never
+			Assert.That(preInvalidations, Has.Count.EqualTo(0));
+
+			//Should receive none invalidation when Cache is configured as never
+			Assert.That(invalidations, Has.Count.EqualTo(0));
+		}
+
+		[Test]
+		public void QueryCache_ThrowsException()
```

Comment:
The test fails for criteria and sql queries as they don't use `QueryLoader`. Here we need to add a test for `CreateCriteria` and `CreateSQLQuery` methods and add a check for them.

Reply:
For native sql I could throw exception in CustomLoader but i didnot use CreateCriteria api where is the best place to put exception

Reply:
Idealy this logic should be in `Loader` so that all share the same check logic. We can achieve this by adding a `QuerySpaces` property to `Loader` which should be overriden by the non abstract subclasses. For `QueryLoader` the property should look like this:
```
public override ISet<string> QuerySpaces => _queryTranslator.QuerySpaces;
```
`CustomLoader` and `CriteriaLoader` have already this property. Then we could modify the `bool IsCacheable(QueryParameters queryParameters)` method of `Loader` to check whether the query is cacheable.

Reply:
ok